File: src/sdk/python/appmesh/tcp_messages.py
"""TCP message classes for HTTP-like communication."""
__all__ = ["RequestMessage", "ResponseMessage"]

# Standard library imports
from dataclasses import dataclass, field
from typing import Any, Dict, Type, get_type_hints

# Third-party imports
import msgpack


@dataclass
class RequestMessage:
    """TCP request message for HTTP-like communication."""

    uuid: str = ""
    request_uri: str = ""
    http_method: str = ""
    client_addr: str = ""
    body: bytes = b""
    headers: Dict[str, str] = field(default_factory=dict)
    query: Dict[str, str] = field(default_factory=dict)

    def serialize(self) -> bytes:
        """Serialize request message to bytes."""
        return msgpack.packb(self.__dict__, use_bin_type=True)


@dataclass
class ResponseMessage:
    """TCP response message for HTTP-like communication."""

    uuid: str = ""
    request_uri: str = ""
    http_status: int = 0
    body_msg_type: str = ""
    body: bytes = b""
    headers: Dict[str, str] = field(default_factory=dict)

    @classmethod
    def from_bytes(cls, buf: bytes) -> "ResponseMessage":
        """Deserialize TCP msgpack buffer with proper type conversion."""
        data = msgpack.unpackb(buf, raw=False)
        hints = get_type_hints(cls)
        instance = cls()

        for key, value in data.items():
            if key in hints:
                setattr(instance, key, cls._convert_type(value, hints[key]))

        return instance

    @staticmethod
    def _convert_type(value: Any, expected_type: Type) -> Any:
        """Convert value to expected type."""
        if value is None:
            return {
                str: "",
                bytes: b"",
                int: 0,
            }.get(expected_type, None)

        if expected_type is str:
            return value.decode("utf-8", errors="replace") if isinstance(value, bytes) else str(value)

        if expected_type is bytes:
            return value.encode("utf-8") if isinstance(value, str) else value

        if expected_type is int:
            return int(value or 0)

        return value

File: src/sdk/rust/src/tcp_messages.rs
// tcp_messages.rs

use serde::{Deserialize, Serialize};
use std::collections::HashMap;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RequestMessage {
    #[serde(default)]
    pub uuid: String,

    #[serde(default)]
    pub request_uri: String,

    #[serde(default)]
    pub http_method: String,

    #[serde(default)]
    pub client_addr: String,

    #[serde(default, with = "serde_bytes")]
    pub body: Vec<u8>,

    #[serde(default)]
    pub headers: HashMap<String, String>,

    #[serde(default)]
    pub query: HashMap<String, String>,
}

impl Default for RequestMessage {
    fn default() -> Self {
        Self {
            uuid: String::new(),
            request_uri: String::new(),
            http_method: String::new(),
            client_addr: String::new(),
            body: Vec::new(),
            headers: HashMap::new(),
            query: HashMap::new(),
        }
    }
}

impl RequestMessage {
    pub fn new() -> Self {
        Self::default()
    }

    /// Serialize to MessagePack format with struct-as-map encoding
    /// This matches Python's msgpack.packb(self.__dict__, use_bin_type=True)
    pub fn serialize(&self) -> Result<Vec<u8>, rmp_serde::encode::Error> {
        let mut buf = Vec::new();
        {
            // Use struct_map mode to serialize as a map with string keys
            // This matches Python's dictionary serialization
            let mut serializer = rmp_serde::Serializer::new(&mut buf).with_struct_map();
            use serde::Serialize;
            Serialize::serialize(self, &mut serializer)?;
        }
        Ok(buf)
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ResponseMessage {
    #[serde(default)]
    pub uuid: String,

    #[serde(default)]
    pub request_uri: String,

    #[serde(default)]
    pub http_status: i32,

    #[serde(default)]
    pub body_msg_type: String,

    #[serde(default, with = "serde_bytes")]
    pub body: Vec<u8>,

    #[serde(default)]
    pub headers: HashMap<String, String>,
}

impl Default for ResponseMessage {
    fn default() -> Self {
        Self {
            uuid: String::new(),
            request_uri: String::new(),
            http_status: 0,
            body_msg_type: String::new(),
            body: Vec::new(),
            headers: HashMap::new(),
        }
    }
}

impl ResponseMessage {
    pub fn new() -> Self {
        Self::default()
    }

    /// Deserialize from MessagePack format
    pub fn deserialize(buf: &[u8]) -> Result<Self, rmp_serde::decode::Error> {
        rmp_serde::from_slice(buf)
    }
}
